### PR TITLE
Add allowed xblocks which an olx may contain but currently not supported

### PR DESCRIPTION
**Pull request by awaisdar001**
Currently, there are many `xblocks` that a course Olx may contain but they are not supported by `olxcleaner`. This PR adds a robust way to bypass that. 

`olxcleaner.validate` API now allows you to provide a list of supported xblocks that may and may not be part of this library already. e.g. 
```
    allowed_xblocks = [
        "recommender",
        "edx_sga",
        "crowdsourcehinter",
        "done",
        "word_cloud",
        "wiki"
    ]
    course, errorstore, url_names = validate("testcourses/testcourse11", steps=1, allowed_xblocks=allowed_xblocks)
```

Providing a list of xblock as `allowed_xblocks` the library will not throw a `UnexpectedTag` error. 

```
"A <recommender> tag was unexpectedly found inside the <vertical url_name='myvertical6' display_name='Hello'> tag")
"A <wiki> tag was unexpectedly found inside the <course display_name='[Test] RecommenderXBlock'> tag")
"A <chapter> tag was unexpectedly found inside the <vertical> tag"
```

FYI -- @pdpinch @jolyonb

https://openedx.atlassian.net/browse/TNL-8147

### CODE REVIEW

**Comment by jolyonb** [comment]
@awaisdar001 Makes me so happy to see somebody pick this up! Thank you!

Do you have a reviewer in mind, or would you like my input too?

**Comment by awaisdar001** [comment]
> @awaisdar001 Makes me so happy to see somebody pick this up! Thank you!
> 
> Do you have a reviewer in mind, or would you like my input too?

@jolyonb sure, I am happy if you can provide some input here.

File: olxcleaner/loader/xml.py [hunk 1]
```diff
@@ -6,37 +6,32 @@
...
+
+
+def load_course(directory, filename, errorstore, allowed_xblocks=None):
     """
     Loads a course, given a filename for the appropriate course.xml file.
 
     :param directory: Path for course.xml (or equivalent)
     :param filename: Filename for course.xml (or equivalent)
     :param errorstore: ErrorStore object to store errors
+    :param allowed_xblocks: ErrorStore object to store errors
```
Comment (asadazam93) [inline olxcleaner/loader/xml.py:34]: Incorrect parameter description

File: olxcleaner/loader/xml.py [hunk 2]
```diff
@@ -80,6 +76,7 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
     :param errorstore: An ErrorStore object that is collecting errors
     :param htmlfiles: A dictionary of XML filenames (value) that reference a given HTML filename (key)
     :param pointer: True if we've arrived at this node due to a pointer tag
+    :param allowed_xblocks: True if we've arrived at this node due to a pointer tag
```
Comment (asadazam93) [inline olxcleaner/loader/xml.py:79]: Incorrect parameter description

File: olxcleaner/exceptions.py [hunk 3]
```diff
@@ -59,3 +61,7 @@ def name(self):
     @property
     def about(self):  # pragma: no cover
         return type(self).__doc__
+
+
+class ClassDoesNotExist(ValueError):
+    """Error raised when we support an xblock but it's support is not available in library."""
```
Comment (asadazam93) [inline olxcleaner/exceptions.py:67]: Error raised when an xblock is allowed but not supported by the library

File: olxcleaner/loader/xml.py [hunk 4]
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
...
+def check_tag_is_allowed(tag, allowed_children, allowed_xblocks):
+    """
+    Checks if a tag is valid or not.
+
+    :param tag: Tag name of the lxml element
+    :param allowed_children: List of xblocks which this tag can have
+    :param allowed_xblocks: List of all allowed xblocks which may or may not be supported currently.
+    """
+    tag_is_allowed = allowed_xblocks is not None and tag in allowed_xblocks
+    if allowed_xblocks is None:
```
Comment (asadazam93) [inline olxcleaner/loader/xml.py:270]: If we check this condition first there will be no need to check for empty in the above line.
```
if allowed_xblocks is None:
    return tag in allowed_children
else:
    return tag in allowed_xblocks
```

File: olxcleaner/loader/xml.py [hunk 5]
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
...
+                    UnexpectedTag(filename, tag=child_tag, edxobj=edxobj)
+                )
+
+
+def check_tag_is_allowed(tag, allowed_children, allowed_xblocks):
+    """
+    Checks if a tag is valid or not.
+
+    :param tag: Tag name of the lxml element
+    :param allowed_children: List of xblocks which this tag can have
```
Comment (asadazam93) [inline olxcleaner/loader/xml.py:266]: List of children/xblock allowed by the olxcleaner

File: olxcleaner/loader/xml.py [hunk 6]
```diff
@@ -239,15 +237,36 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
...
+    :param allowed_xblocks: List of all allowed xblocks which may or may not be supported currently.
```
Comment (asadazam93) [inline olxcleaner/loader/xml.py:267]: List of Xblocks that are allowed but may not be supported by the olxcleaner for validation

File: olxcleaner/loader/xml.py (hunk 4 above)
Comment (awaisdar001) [reply olxcleaner/loader/xml.py:269]: Done.

File: olxcleaner/loader/xml.py (hunk 1 above)
Comment (awaisdar001) [reply olxcleaner/loader/xml.py]: Done.

**Comment by awaisdar001** [comment]
@asadazam93 addressed your comments.

File: olxcleaner/loader/xml.py [hunk 7]
```diff
@@ -239,15 +237,35 @@ def read_course(edxobj, node, directory, filename, errorstore, htmlfiles, pointe
...
+    """
+    Checks if a tag is valid or not.
+
+    :param tag: Tag name of the lxml element
+    :param allowed_children: List of children/xblocks allowed by the olxcleaner
+    :param allowed_xblocks: List of Xblocks that are allowed but may not be supported by the olxcleaner for validation
+    """
+    if allowed_xblocks is None:
+        return tag in allowed_children
+    return tag in allowed_xblocks
```
Comment (jolyonb) [inline olxcleaner/loader/xml.py:271]: This logic doesn't seem quite right to me. I think you want `return tag in allowed_xblocks or tag in allowed_children`.

**Review by asadazam93 — APPROVED** [review]

File: olxcleaner/loader/xml.py (hunk 7 above)
Comment (awaisdar001) [reply olxcleaner/loader/xml.py:271]: as `allowed_xblocks` can be None so the above-mentioned code can break.

**Comment by jolyonb** [comment]
I'm a little concerned that the approach you're taking here allows errors to slip through. For example, if you have `vertical` in `allowed_xblocks`, then your code will allow `vertical` to be a child of any other xblock, including, for example, `vertical`. I suspect this can be fixed by filtering out classes that are already validated from `allowed_xblocks`.

File: olxcleaner/loader/xml.py (hunk 7 above)
Comment (jolyonb) [reply olxcleaner/loader/xml.py:271]: This is intended to replace line 271, not line 269-270.

Comment (awaisdar001) [reply olxcleaner/loader/xml.py:271]: Ah I see, I understand.

**Comment by awaisdar001** [comment]
> I'm a little concerned that the approach you're taking here allows errors to slip through. For example, if you have `vertical` in `allowed_xblocks`, then your code will allow `vertical` to be a child of any other xblock, including, for example, `vertical`. I suspect this can be fixed by filtering out classes that are already validated from `allowed_xblocks`.

Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?

**Comment by jolyonb** [comment]
> Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?

Exactly.

**Comment by awaisdar001** [comment]
> > Thanks for the input @jolyonb -- Do you mean to filter out values from `allowed_xblocks` that already exist in `EdxObject.__subclasses__()`?
> 
> Exactly.

👍 I'll take a look.

**Comment by awaisdar001** [comment]
@jolyonb I have addressed your comments, can you take a look at the following commit?
https://github.com/openedx/olxcleaner/pull/1/commits/fd46d1d4a3e2c8834fe50af7f0b1d80d6b42940f

**Comment by jolyonb** [comment]
Looks good to me now! 👍

**Comment by awaisdar001** [comment]
> Looks good to me now! 👍

Thanks @jolyonb 🎆

**Comment by awaisdar001** [comment]
🎆 first Pr in this fork 🎆

**Pull request merged**